satellite/metainfo/metainfo.go

```diff
@@ -1636,6 +1636,20 @@ func (endpoint *Endpoint) BeginDeleteSegment(ctx context.Context, req *pb.Segmen
 		}
 	}
 
+	// moved from FinishDeleteSegment to avoid inconsistency if someone will not
+	// call FinishDeleteSegment on uplink side
+	for _, piece := range pointer.GetRemote().GetRemotePieces() {
+		_, err := endpoint.containment.Delete(ctx, piece.NodeId)
```

Comment:
Shouldn't containment deletion also pass in the piece id? Otherwise nodes would be removed from containment all the time?

Reply:
Alternatively, don't remove from containment, and let containment handling check whether the piece still needs to be checked?

Reply:
> Shouldn't containment deletion also pass in the piece id? Otherwise nodes would be removed from containment all the time?

That was the logic before new metainfo calls. 

> Alternatively, don't remove from containment, and let containment handling check whether the piece still needs to be checked?

Are you sure containment will handle this correctly? It would be cleaner to remove such custom logic from metainfo.

Reply:
> Are you sure containment will handle this correctly?

At the moment, I'm not sure. I think it's worth making this into a separate PR, so that metainfo doesn't need to do these checks.

Reply:
What if we modified `containment.Get()` to check metainfo before returning the node? It would be adding some complexity to containment mode, but if we don't do that we would need to make sure we are properly updating containment mode in several places (not only delete segment).

Reply:
We could add a function in containment mode to conditionally delete a node from containment if its piece matches the one being deleted, but there could be too many places to maintain this logic.

Reply:
Tracked this in https://storjlabs.atlassian.net/browse/V3-2686

Reply:
> ... make the containment DB dependent on the metainfo service ...

This would actually imply we are missing a containment service, because there's more logic missing.

> I think instead, since the audit service already has access to metainfo and containment, and it is the primary service interacting with containment, deletion should be done from audit.Reverify(). 

Yes, that could work as well.

> Another solution would be to add a conditional delete function to containment that will delete only if the path + piece that a node was contained for matches arguments passed in. The downside is that we would have to call it whenever we update pointers, which would be more difficult to maintain long term.

As you mentioned, this indeed would be fragile.

Reply:
> This would actually imply we are missing a containment service, because there's more logic missing.
We are missing a containment service. What do you think the better option would be between adding a service + extra logic around Get/Increment vs. dealing with deletion from audit? I think it would be better to just add the logic in audit for now since it would be a lot less code, but it probably loses some of the advantages of adding a new service.

Reply:
Yeah, adding the code into audit would be fine solution for now.